### PR TITLE
Surface Smoothing

### PR DESCRIPTION
**Pull request by Edouard2laire**
Hello, 

This PR has three objectives: 
- (1) expose the smoothing function so it can be called more  easily from other scripts
- (2) Fix a bug on the computation of the mean distance of the vertex (that was using the binary vertCon as coordinate instead of the vertex coordinate) 
- (3) Smooth each connected region of the surface separately (useful when the mesh density is not the same in all regions)

(2)  
The following figure shows the impact of the error on the distance calculation by plotting the histogram of the distance between two connected vertex and the average used for calculation. For the current version,  the distances can only take 4 values: 0, sqrt(1), sqrt(2), or sqrt(3). Most of them are 0 and not shown. Surprisingly the result ended up being not that far ( 2.29 mm vs. 5.80 mm) -- Note it seems that the more dense the mesh is, the larger the issue is. 

untitled

Graph created using the following code: 

```matlab
    figure;
    ax1 = subplot(121);hold on;
    Vertices = SurfaceMat.VertConn;
    [vi,vj] = find(SurfaceMat.VertConn);
    tmp = sqrt((Vertices(vi,1) - Vertices(vj,1)).^2 + (Vertices(vi,2) - Vertices(vj,2)).^2 + (Vertices(vi,3) - Vertices(vj,3)).^2);
    histogram( tmp (tmp > 0 ));
    xline(mean(tmp),'-',sprintf('Mean Distance %.2f mm', 1000* full(mean(tmp))))
    xlim([-0.1 max(xlim(gca))])
    title('Vertices = SurfaceMat.VertConn;')


    ax2 = subplot(122); hold on;
    Vertices = SurfaceMat.Vertices;
    tmp = sqrt((Vertices(vi,1) - Vertices(vj,1)).^2 + (Vertices(vi,2) - Vertices(vj,2)).^2 + (Vertices(vi,3) - Vertices(vj,3)).^2);
    histogram( tmp (tmp > 0 ));
    xline(mean(tmp),'-',sprintf('Mean Distance %.2f mm', 1000* mean(tmp)))
    title('Vertices = SurfaceMat.Vertices;')

```
Note: this might be linked to the issue [JonathanC](https://neuroimage.usc.edu/forums/u/JonathanC) was trying to explain on the forum https://neuroimage.usc.edu/forums/t/spatial-smoothing-process-modification/11157/13 

(3)  Actually, I think it would be nice to find a solution that is not working based on the number of vertexes to smooth as this is not 100% constant across the cortex** and can vary even more if the surface used contains an additional surface with more vertices such as cerebellum or hippocampus. For example, when you merge the mid-surface with the hippocampus, the distribution of distance is as follows: 

untitled2

where regions 1 and 2 are from the hippocampus, and regions 3-4 are from the cortex. See the following figure:
So using the same number of vertex for all the regions will lead to a wrong smoothing in some regions. (figure created using: https://gist.github.com/Edouard2laire/70f5124b305432440db348715126d1a1)

untitled3

Here is the results of different smoothing: 

Screenshot from 2023-09-20 18-31-19

From left to right: the original map, the current version, the new version with a fixed FHMW, and the version with an FHWM per region. In this case, surfstat was called with 5.671 edges, 5.641 edges, 2.199 edges, and 2.18 edges. 

The difference between the last two versions is mainly in the hippocampus. 

Regards, 
Edouard

---
## Results with the code in this PR
Visual comparison of the spatial smoothing methods (surface 15k vertices), for two values of FWHM: 3 mm and 30 mm.

For both FWHM, the `Fix-` and `Adaptive-FWHM` methods (in old process) lead to similar results as the mean distance is similar in both hemispheres. These results are also similar to the ones obtained with the `Geodesic distance in edges` method (in new process).

As mentioned by @Edouard2laire, with this FWHM and surface, the bug implementation still gives good results for small FWHM in this surface size. However the bug has a big impact when smoothing with larger FWHM values (in this example surface) or smoothing surfaces with different mean distance (to this example surface). 

image

image

### CODE REVIEW

**Comment by Edouard2laire** [comment]
This is ready to reviewed :) Let me know if you have any comments :) 

Do you know why https://github.com/brainstorm-tools/brainstorm3/blob/master/toolbox/process/deprecated/process_ssmooth.m why deprecated ?

**Comment by tmedani** [comment]
I ll review it asap

Thanks

**Comment by Edouard2laire** [comment]
Thank you. 

I was thinking, maybe in order to facilitate, the review that we should decide first on what we will recommend for the smoothing and only implement those options. (probably merging the two open requests)

Here is what I was thinking: 
- For existing study: continue using the current code ( we should add a warning to indicate the actual FWHM used in the code)

- For new study if the mesh is homogeneous: 
-- > (1)  Use smoothing using the actual geodesic distance between the vertex (implemented in #646 
---> (2) Using surf stat using the corrected code for the average distance between vertex (2023_fixed in the code)
---> (3) Using surf stat using the corrected code for the average distance between vertex  using a different FWHM for each region (2023_adptive in the code)

- For  new studies if the mesh is not homogeneous: 
-- > (1)  Use smoothing using the actual geodesic distance between the vertex (implemented in #646 
---> (2) Using surf stat using the corrected code for the average distance between vertex  using a different FWHM for each region (2023_adptive in the code)
---> (3) Using surf stat using the corrected code for the average distance between vertex (2023_fixed in the code)


Although I like the adaptative FWHM, I don't think it is necessary to put it in the interface as the smoothing in #646 might be a better way to handle that. So I would keep only 3 options: 2 using Surfstat (with or without the bug) and one version using our implementation.

**Comment by Edouard2laire** [comment]
any update? :)

**Comment by Edouard2laire** [comment]
any updates ?

**Comment by rcassani** [comment]
Hi @Edouard2laire, apologies for the long delay in this PR.

Just made few aesthetic changes in the process.

---

> (2) Fix a bug on the computation of the mean distance of the vertex (that was using the binary vertConn as coordinate instead of the vertex coordinate)

It seems, it went under the radar as the estimated distances is not that bad for the 15k-vertex surface.

---
> Do you know why https://github.com/brainstorm-tools/brainstorm3/blob/master/toolbox/process/deprecated/process_ssmooth.m was deprecated ?

Not sure, this happened in Feb 2016.

--- 

Lastly, about the different options. Another options would be to:

* Move the current spatial smoothing process (without changes) to deprecated
* Bring the code from the #646 to this PR, and have a new spatial smoothing process with the options:
  * Distance metric:  ○ Euclidean, ○ geodesic
  * FWHM value [mm]
  * Checkbox: Use adaptive approach

In this sense, spatial smoothing for old scripts will still available in the deprecated process. The new process would not have that problem.

What do you think?
@tmedani, any other idea?

**Comment by tmedani** [comment]
@rcassani, thanks for following up on that and confirming the issue that @Edouard2laire found and solved.
I agree to remove it from the GUI and keep it "deprecated" for scripting and for reproducibility purposes.

**Comment by Edouard2laire** [comment]
> Hi @Edouard2laire, apologies for the long delay in this PR.

No problem. I know we are all busy. Let me know if there is anything I can do to help make future PR reviews easier for you. 

> Just made few aesthetic changes in the process.
> 
> > (2) Fix a bug on the computation of the mean distance of the vertex (that was using the binary vertConn as coordinate instead of the vertex coordinate)
> 
> It seems, it went under the radar as the estimated distances is not that bad for the 15k-vertex surface.
> 
> > Do you know why https://github.com/brainstorm-tools/brainstorm3/blob/master/toolbox/process/deprecated/process_ssmooth.m was deprecated ?
> 
> Not sure, this happened in Feb 2016.

From what I understood in a forum post I cannot find back, the function to calculate geodesic distance was made by hand by Francois and was quite slow. I guess Matlab has some nice trick to get the distance in an efficient manner making the process now easy to use. 



> 
> Lastly, about the different options. Another options would be to:
> 
> * Move the current spatial smoothing process (without changes) to deprecated
> * Bring the code from the [Distance computation (euclidean / geodesic) #646](https://github.com/brainstorm-tools/brainstorm3/pull/646) to this PR, and have a new spatial smoothing process with the options:
>   
>   * Distance metric:  ○ Euclidean, ○ geodesic
>   * FWHM value [mm]
>   * Checkbox: Use adaptive approach
> 
> In this sense, spatial smoothing for old scripts will still available in the deprecated process. The new process would not have that problem.
> 
> What do you think? @tmedani, any other idea?


> * Move the current spatial smoothing process (without changes) to deprecated


I would keep the changes. Especially since the previous script will still work using the same option. I added a warning explaining the issue 

image


> * Bring the code from the [Distance computation (euclidean / geodesic) #646](https://github.com/brainstorm-tools/brainstorm3/pull/646) to this PR, and have a new spatial smoothing process with the options:
>   
>   * Distance metric:  ○ Euclidean, ○ geodesic
>   * FWHM value [mm]
>   * Checkbox: Use adaptive approach


I am not sure I understand what the Euclidean option. If both are not using surfstat, then I would only propose the geodesic option since both will have similar computation time but the latter is more accurate. 

My preference would now be: 

>   
>   * Method :  ○ Surfstat (Before 2023 ) , ○ Surfstat (After 2023)  ○ geodesic (recommended)
>   * FWHM value [mm]
>   * Checkbox: Use adaptive approach


My idea with that is to allow a smooth transition for the user as they will see the option they used to know  Surfstat (Before 2023 ) and have the choice of with method they want to replace it with  Surfstat (After 2023) or geodesic (recommended)

**Comment by Edouard2laire** [comment]
Thank you for the changes. it seems to make sense to me. 

One thing I am afraid of is that users might be confused when they launch the smoothing process as to why the interface changed and what option they should choose.  Also, we should maybe add a message saying they shouldn't mix result smoothed with this process and the previous one as the level of smoothing would be different and might biased their results (esp if they are looking at thing like source extent, I guess)

**Comment by rcassani** [comment]
@Edouard2laire, the PR should be ready to merge, you may want to run a quick with adaptive FWHM for disconnected surfaces with very distinct mean distance.

These are the changes for the last series of commits:
1. Process `Smooth surface` (this is the old one)
    * Three options: Old (with bug), fix FWHM, adaptive FWHM
    * This process will be moved to deprecated
    * Bug is more noticeable for large FWHM (see images below)
  
2. Process `Smooth surface [2024]` (new one)
    * Two options: Geodesic distance (mm), Geodesic distance (in edges) with adaptive FWHM

3. Others
    * Moved finding of connected regions to `process_ssmooth`
    * Fix issue with output file comment for the case of using absolute values on sources

## Figures
Visual comparison of the spatial smoothing methods (surface 15k vertices), for two values of FWHM: 3 mm and 30 mm.

For both FWHM, the `Fix-` and `Adaptive-FWHM` methods (in old process) lead to similar results as the mean distance is similar in both hemispheres. These results are also similar to the ones obtained with the `Geodesic distance in edges` method (in new process).

As mentioned by @Edouard2laire, with this FWHM and surface, the bug implementation still gives good results for small FWHM in this surface size. However the bug has a big impact when smoothing with larger FWHM values (in this example surface) or smoothing surfaces with different mean distance (to this example surface). 

image

image

**Comment by Edouard2laire** [comment]
Sorry, I didn't see the notification of your message. 


I made two small changes: 
- expose the compute function in the smooth process to have easier access to external processes
- Tried to uniformize the file name in the db and history

File: toolbox/process/functions/process_ssmooth.m [hunk 1]
```diff
@@ -113,9 +112,40 @@
...
+            msgInfo = sprintf('Spatial smoothing using %1.2f mm kernel calculating distance using geodesic distance', FWHM*1000);
+        case 'geodesic_edge'
+            msgInfo = sprintf('Spatial smoothing using %1.2f mm kernel calculating distance using edge path length distance', FWHM*1000);
+    end
+    errInfo = '';
+
     % ===== COMPUTE SMOOTHING OPERATOR =====
     % Get existing interpolation for this surface
-    Signature = sprintf('ssmooth(%1.2f,%s):%s', round(FWHM*1000), Method, FileMat.SurfaceFile);
+    Signature = sprintf('ssmooth(%1.2f,%s):%s', round(FWHM*1000), Method, SurfaceMat.Comment);
```
[the remark this answers is not in the document]
Comment (rcassani) [reply toolbox/process/functions/process_ssmooth.m]: Should be solved by [0235f31](https://github.com/brainstorm-tools/brainstorm3/pull/645/commits/0235f319596d1f69d9b09ce0e6836b07555f590a)

**Comment by Edouard2laire** [comment]
This is working, but makes it harder to call the function from outside as it means duplicating the code to get the Winterp.  Maybe we could save it on the cortical surface file. 

Or actually. maybe we don’t actually need to save it and it would make more sense to only save the geodesic distance inside the surface.  As this is the part that needs the most calculation and that can be reused for multiple purpose.  
esp computing the kernel from the distance is relatively cheap.

**Comment by Edouard2laire** [comment]
What about something like that?

I would have preferred to save the distance * but it's quite painful to have the filename of the surface inside bst_tess_distance

* a bit like what is done tess_interp_mri

**Comment by rcassani** [comment]
I don't think saving the distance matrix is the way to go:
* When the distance is not saved yet: 
  * It takes quite long to save it in the surface file
  * The size of the surface files goes from 17MB to 870MB (for 15k vertices)
* When the distance is already saved:
  * The surface file is read completely twice: in `compute()` and in `bst_tess_distance()`, and it's slow

Keeping the smoothing kernel in RAM seems like a good approach. If we want to store it, then it could be saved in HDD (as sparse in the field `tess2tess`).

**Comment by Edouard2laire** [comment]
ok. 

I'll revert the changes to save the kernel in memory as before.

**Comment by Edouard2laire** [comment]
ok, something like that should be working.

**Comment by rcassani** [comment]
@Edouard2laire, the PR seems to be ready to be merged.
Let me know if you want to add something else.

- The figures (and their text) of the results obtained with this PR code are now part of the PR description

**Comment by Edouard2laire** [comment]
it's good for me. Thanks a lot for your help in this PR :)

**Pull request merged**